### PR TITLE
Add egbuilder

### PR DESCRIPTION
**Pull request by suchen-sci**

### CODE REVIEW

File: cmd/builder/gen/controllers.go [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+ * limitations under the License.
+ */
+
+package gen
+
+import (
+	"fmt"
+	"strings"
+
+	j "github.com/dave/jennifer/jen"
```
Comment (localvar) [inline cmd/builder/gen/controllers.go:24]: this is fine, but I think using 'text/template' for this task is better.

File: cmd/builder/command/init.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+	}
+	fmt.Println("generate filters success")
+
+	err = initConfig.GenControllers(cwd)
+	if err != nil {
+		utils.ExitWithError(err)
+	}
+	fmt.Println("generate controllers success")
+
+	err = initConfig.GenRegistry(cwd)
```
Comment (localvar) [inline cmd/builder/command/init.go:115]: will all controllers and filters be added to registry even some of them failed to generate?

File: cmd/builder/gen/controllers.go (hunk 1 above)
Comment (suchen-sci) [reply cmd/builder/gen/controllers.go:24]: `jennifer` can do some checks. And can split the generated files into pieces. Emmm, it's a trade off choice.

File: cmd/builder/command/init.go (hunk 2 above)
Comment (suchen-sci) [reply cmd/builder/command/init.go:115]: No, if generation of file failed, the process will exit.

File: doc/egbuilder.md [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+- module: github.com/other/repo
+
+# output: path of output file.
+output: "./easegress-server"
+
+# raceDetector: "-race" flag for go build
+raceDetector: false
+
+# skipBuild: if true, not build, just generate temp directory and files. 
+# still clean up if skipCleanUp is false.
```
[the remark this answers is not in the document]
Comment (suchen-sci) [reply doc/egbuilder.md]: all updated, please check!

**Pull request merged**